tablets.go

```diff
@@ -0,0 +1,207 @@
+			l1 = mid + 1
+		} else {
+			r1 = mid
+		}
+	}
+	start := l1
+
+	if start > l && t[start-1].LastToken() > tablet.FirstToken() {
+		start = start - 1
+	}
+
+	// find last overlaping range
+	for l2 < r2 {
+		mid := (l2 + r2) / 2
+		if t[mid].LastToken() < tablet.LastToken() {
+			l2 = mid + 1
+		} else {
+			r2 = mid
+		}
+	}
+	end := l2
+	if end < r && t[end].FirstToken() >= tablet.LastToken() {
+		end = end - 1
+	}
+	if end == len(t) {
+		end = end - 1
+	}
+
+	updated_tablets := t
+	if start <= end {
+		// Delete elements from index start to end
+		updated_tablets = append(t[:start], t[end+1:]...)
+	}
+	// Insert tablet element at index start
+	t = append(updated_tablets[:start], append([]*TabletInfo{tablet}, updated_tablets[start:]...)...)
+	return t
+}
+
+// Remove all tablets that have given host as a replica
+func (t TabletInfoList) removeTabletsWithHostFromTabletsList(host *HostInfo) TabletInfoList {
+	filteredTablets := make([]*TabletInfo, 0, len(t)) // Preallocate for efficiency
+
+	for _, tablet := range t {
+		// Check if any replica matches the given host ID
+		shouldExclude := false
+		for _, replica := range tablet.replicas {
+			if replica.hostId.String() == host.HostID() {
+				shouldExclude = true
+				break
+			}
+		}
+		if !shouldExclude {
+			filteredTablets = append(filteredTablets, tablet)
+		}
+	}
+
+	t = filteredTablets
+	return t
+}
+
+func (t TabletInfoList) removeTabletsWithKeyspaceFromTabletsList(keyspace string) TabletInfoList {
+	filteredTablets := make([]*TabletInfo, 0, len(t))
+
+	for _, tablet := range t {
+		if tablet.keyspaceName != keyspace {
+			filteredTablets = append(filteredTablets, tablet)
+		}
+	}
+
+	t = filteredTablets
+	return t
+}
+
+func (t TabletInfoList) removeTabletsWithTableFromTabletsList(keyspace string, table string) TabletInfoList {
+	filteredTablets := make([]*TabletInfo, 0, len(t))
+
+	for _, tablet := range t {
+		if !(tablet.keyspaceName == keyspace && tablet.tableName == table) {
+			filteredTablets = append(filteredTablets, tablet)
+		}
+	}
+
+	t = filteredTablets
+	return t
+}
+
+// Search for place in tablets table for token starting from index l to index r
+func (t TabletInfoList) findTabletForToken(token Token, l int, r int) *TabletInfo {
+	for l < r {
+		var m int
+		if r*l > 0 {
+			m = l + (r-l)/2
+		} else {
+			m = (r + l) / 2
+		}
+		if int64Token(t[m].LastToken()).Less(token) {
+			l = m + 1
+		} else {
+			r = m
+		}
+	}
+
+	return t[l]
+}
+
+// cowTabletList implements a copy on write tablet list, its equivalent type is TabletInfoList
+type cowTabletList struct {
+	list atomic.Value
+	mu   sync.Mutex
+}
+
+func (c *cowTabletList) get() TabletInfoList {
+	l, ok := c.list.Load().(TabletInfoList)
+	if !ok {
+		return nil
+	}
+	return l
+}
+
+func (c *cowTabletList) set(tablets TabletInfoList) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	n := len(tablets)
+	t := make(TabletInfoList, n)
+	for i := 0; i < n; i++ {
+		t[i] = tablets[i]
+	}
```

Comment:
Shallow copy of list of pointers is not exactly safe.
If you really want to make it safe you need to implement proper `.Clone` method for both `TabletInfo` and `TabletInfoList`.

It needs to be done only if we are planing on manipulating `TabletInfo.replicas`, otherwise simple `c.list.Store` will be enough:
```
	c.mu.Lock()
	c.list.Store(t)
	c.mu.Unlock()
```